### PR TITLE
Protocol version

### PR DESCRIPTION
**Pull request by kaikai1024**
### Description of the Change

Use new interface:

* getProtocolVersion
* setProtocolVersion

old interfaces still in use.

> @notice Deprecated.

### Applicable Issues

#492

### CODE REVIEW

**Comment by ouwenkg** [comment]
ci...

**Comment by kaikai1024** [comment]
> ci...

ok. Need to update the genesis.

**Comment by kaikai1024** [comment]
There's a problem: It's required to amend the code of version manager when the chain needs a upgrade to the next version.
I'll check if i can skip the upgrade

**Comment by kaikai1024** [comment]
One version to prepare the upgrade.
And will update the doc to show the info.

**Comment by ouwenkg** [comment]
> There's a problem: It's required to amend the code of version manager when the chain needs a upgrade to the next version.
> I'll check if i can skip the upgrade

We need a script tool or operation guide docs.

**Comment by kaikai1024** [comment]
> We need a script tool or operation guide docs.

yes. but i want the upgrade is a choice for users. maybe a guide doc or use `cita-cli`, i don't want to depend the txtool like `amend_system_contracts`. 
by the way, `amend_system_contracts` is useless.

File: scripts/contracts/src/system/VersionManager.sol [hunk 1]
```diff
@@ -37,11 +37,28 @@ contract VersionManager is IVersionManager, ReservedAddrPublic {
...
 
-    function getVersion()
+    /// @notice Deprecated. Check the setProtocolVersion
+    function setVersion(uint32 _version)
+        public
+        onlyAdmin
+    {
+        setProtocolVersion(_version);
+    }
+
```
Comment (ouwenkg) [inline scripts/contracts/src/system/VersionManager.sol]: 这样实现不太好吧, 保持原来的 `set_version` 不动, 然后新增加一个 `setProtocolVersion` 接口就好了.

Comment (kaikai1024) [reply scripts/contracts/src/system/VersionManager.sol]: Don't get your point. 
It's just like what you say, old version uses the old interface, new version uses the new interface.

Comment (ouwenkg) [reply scripts/contracts/src/system/VersionManager.sol]: OK, i understand.

**Comment by ouwenkg** [comment]
Please update `contracts.yml` permission section.

**Comment by kaikai1024** [comment]
> Please update `contracts.yml` permission section.

opp, I forget this.

Change the system contracts is a bad idea....
Actually it should change the rpc interface(getSoftwareVersion), but i don't want to trouble sdk or test.

**Comment by ouwenkg** [comment]
> > Please update `contracts.yml` permission section.
> 
> opp, I forget this.
> 
> Change the system contracts is a bad idea....
> Actually it should change the rpc interface(getSoftwareVersion), but i don't want to trouble sdk or test.


> > Please update `contracts.yml` permission section.
> 
> opp, I forget this.
> 
> Change the system contracts is a bad idea....
> Actually it should change the rpc interface(getSoftwareVersion), but i don't want to trouble sdk or test.

No matter which way we choose, trouble sdk or test is inevitable?

**Comment by kaikai1024** [comment]
> No matter which way we choose, trouble sdk or test is inevitable?

Not a level. 
one is a base rpc interface(both sdk and test need to work to fix it.), the other one is just a contract interface call(sdk can skip it and test can add a new case.).

File: cita-executor/core/src/contracts/solc/version_management.rs [hunk 2]
```diff
@@ -30,6 +30,8 @@ use types::ids::BlockId;
 use types::reserved_addresses;
 
 lazy_static! {
+    // Will use `getProtocolVersion` at next version after `v0.25.0`.
+    // And the `getVersion` will be *Deprecated*.
     static ref VERSION_HASH: Vec<u8> = method_tools::encode_to_vec(b"getVersion()");
```
Comment (ouwenkg) [inline cita-executor/core/src/contracts/solc/version_management.rs]: Why next version?

File: scripts/contracts/contracts.yml [hunk 3]
```diff
@@ -223,3 +223,4 @@ PermissionContracts:
         - VersionManager
       functions:
         - 'setVersion(uint32)'
+        - 'setProtocolVersion(uint32)'
```
Comment (ouwenkg) [inline scripts/contracts/contracts.yml]: Please update permission section in genesis module.

File: cita-executor/core/src/contracts/solc/version_management.rs (hunk 2 above)
Comment (kaikai1024) [reply cita-executor/core/src/contracts/solc/version_management.rs]: The new coming version will not enforce user to amend code when upgrade.

File: scripts/contracts/contracts.yml (hunk 3 above)
Comment (kaikai1024) [reply scripts/contracts/contracts.yml]: ok

File: scripts/contracts/contracts.yml [hunk 4]
```diff
@@ -223,3 +223,4 @@ PermissionContracts:
         - VersionManager
       functions:
```
Comment (ouwenkg) [inline scripts/contracts/contracts.yml]: 这里应该有两个 VersionManager 吧

File: cita-executor/core/src/contracts/solc/version_management.rs (hunk 2 above)
Comment (ouwenkg) [reply cita-executor/core/src/contracts/solc/version_management.rs]: Ok

File: scripts/contracts/contracts.yml (hunk 4 above)
Comment (kaikai1024) [reply scripts/contracts/contracts.yml]: yes

**Pull request merged**